f/common_logic/data_validation.py

```diff
@@ -0,0 +1,152 @@
+logger = logging.getLogger(__name__)
+
+
+def detect_structured_data_type(file_path: str) -> str:
+    """
+    Detects the type of a structured data file, focusing on geospatial and tabular formats.
+
+    Only files with recognized extensions are considered. The function attempts to
+    parse the file to confirm it matches the expected structure. If the file type is
+    unsupported or the content does not validate, 'unsupported' is returned.
+
+    This function is intended to identify structured datasets used in environmental,
+    geographic, and tabular data pipelines. It supports:
+    - Geospatial formats: geojson, kml, gpx
+    - Tabular formats: csv, xls, xlsx
+    - General structured formats: json
+
+    TODO: Add support for ESRI shapefiles
+
+    Parameters
+    ----------
+    file_path : str
+        Path to the file to inspect.
+
+    Returns
+    -------
+    str
+        The validated file type (e.g., "csv", "geojson", etc.), or "unsupported" if
+        the file does not conform to expectations.
+    """
+
+    path = Path(file_path)
+    if not path.exists():
+        raise FileNotFoundError(f"File does not exist: {file_path}")
+
+    # Only allow files with known extensions
+    suffix_map = {
+        ".geojson": "geojson",
+        ".json": "json",
+        ".xlsx": "xlsx",
+        ".xls": "xls",
+        ".gpx": "gpx",
+        ".kml": "kml",
+        ".csv": "csv",
+    }
+
+    suffix = path.suffix.lower()
+
+    logger.debug(f"Checking file {file_path} with suffix {suffix}")
+
+    if suffix not in suffix_map:
+        return "unsupported"
+
+    expected_type = suffix_map[suffix]
+
+    def get_namespace(tag: str) -> str:
+        if tag.startswith("{"):
+            return tag[1:].split("}")[0]
+        return ""
+
+    # Try validators in priority order; first match wins
+    def is_json(path):
+        with path.open() as f:
+            json.load(f)
+        return True
+
+    def is_geojson(path):
+        with path.open() as f:
+            data = json.load(f)
+        return isinstance(data, dict) and data.get("type") == "FeatureCollection"
+
+    def is_excel(path):
+        pd.read_excel(path)
+        return True
+
+    def is_gpx(path):
+        # https://www.topografix.com/gpx/1/1/#SchemaProperties
+        # We don't care about the schema version, just that it's a GPX file.
+        tree = etree.parse(str(path))
+        ns = get_namespace(tree.getroot().tag)
+        return ns.startswith("http://www.topografix.com/GPX")
+
+    def is_kml(path):
+        # https://developers.google.com/kml/documentation/kmlreference
+        # We don't care about the schema version, just that it's a KML file.
+        tree = etree.parse(str(path))
+        ns = get_namespace(tree.getroot().tag)
+        return ns.startswith("http://www.opengis.net/kml")
+
+    def is_csv(path):
+        # CSV detection is a bit more complex than the other formats,
+        # because it's overly permissive.
```

Comment:
Why are we ok using `pd.read_excel` but not `pd.read_csv`?

You mentioned in the PR description not wanting `csv.reader` but I am asking you to justify that position.